### PR TITLE
Fix #4878: Compile error when using destructuring with a splat or expansion in an array

### LINKED ISSUE #4878: Nodes: Compile error when using destructuring with a splat or expansion in an array
```coffee
foo = (list) ->
  [ first, ..., last ] = list
```

results in `TypeError: Cannot read property 'indexOf' of undefined`

That’s the simplest case I could come up with. It’s pretty finicky though, if you reference the values after the destructuring statement it compiles:

```coffee
foo = (list) ->
  [ first, ..., last ] = list
  [ first, last ]
```

However other more complex variations also fail, for instance having that statement inside conditional assignment:

```coffee
foo = (list) ->
  ret =
    if list.length
      [ first, ..., last ] = list
      [ first, last ]
    else
      []
```

removing the `ret = ` part of the last example compiles fine

Not a huge deal, it can be replaced with code that explicitly uses `list.slice -1`, it’s just we had some code that was compiling previously and stopped working with a minor update somewhere along the way

### PR DESCRIPTION
**Pull request by zdenko**
Fixes #4878.

### CODE REVIEW

File: test/assignment.coffee [hunk 1]
```diff
@@ -945,3 +945,66 @@ test "#4673: complex destructured object spread variables", ->
 
   {{g}...} = g: 1
   eq g, 1
+
+test "#4878: Compile error when using destructuring with a splat or expansion in an array", ->
+  eqJS '''
+    foo = (list) ->
+      [first, ..., last] = list
+    ''', '''
```
[the remark this answers is not in the document]
Comment (zdenko) [reply test/assignment.coffee]: This example would compile on `master`.
This PR fixes the issue where destructuring is the last line.
```coffeescript
foo = (list) ->
  [ first, ..., last ] = list
```
But, anyway, you're right. I'll change the tests.

File: test/assignment.coffee [hunk 2]
```diff
@@ -947,64 +947,41 @@ test "#4673: complex destructured object spread variables", ->
...
+  f2 = (list) ->
+    [first..., last] = list
+
+  f3 = (list) ->
+    ([first, ...] = list); first
+
+  f4 = (list) ->
+    ([first, ...rest] = list); rest
+
+  arrayEq f1(arr), ['d']
```
Comment (GeoffreyBooth) [inline test/assignment.coffee]: In 2.1.1, `f1(arr)` returns `[ 'a', 'b', 'c', 'd' ]`

File: test/assignment.coffee [hunk 3]
```diff
@@ -947,64 +947,41 @@ test "#4673: complex destructured object spread variables", ->
...
+  arrayEq f2(arr), ['d']
+  arrayEq f3(arr), 'a'
+  arrayEq f4(arr), ['b', 'c', 'd']
+
+  foo = (list) ->
+    ret =
+      if list.length
```
Comment (GeoffreyBooth) [inline test/assignment.coffee]: I don’t like using `list.length` being nonzero as a proxy for truthiness. If you’re evaluating that `list.length isnt 0`, you should be explicit about it. I think it’s a lot more readable than just `list.length` or other hacks like `!!`.

File: test/assignment.coffee (hunk 2 above)
Comment (zdenko) [reply test/assignment.coffee]: In 2.1.1 this
```coffeescript
f1 = (list) ->
 [first, ..., last] = list
```

compiles into 

```javascript
f1 = function(list) {
  var first, last;
  return first = list[0], last = list[list.length - 1], list;
};
```

So, the returned value is `list`.

In 2.2.0 it's compiled to

```javascript
f1 = function(list) {
  var first, last;
  return [first] = list, [last] = slice.call(list, -1);
};
```

The returned value is `[last]'.

File: test/assignment.coffee (hunk 3 above)
Comment (zdenko) [reply test/assignment.coffee]: Right. I copied the example from #4878.

File: test/assignment.coffee (hunk 2 above)
Comment (GeoffreyBooth) [reply test/assignment.coffee]: Yes, but `[first, ..., last] = list` is shorthand for `return [first, ..., last] = list`, which looks to me like it should return the full array. As in, returning just `[last]` is wrong.

```coffee
do -> [a, b, c] = [1, 2, 3]
# returns [1, 2, 3]
```

Comment (zdenko) [reply test/assignment.coffee]: Yup. Sorry, my mistake. Not enogh coffe in the system :sleepy:

Comment (GeoffreyBooth) [reply test/assignment.coffee]: > Not enough coffee in the system

Isn’t that why you’re here? 😝

**Comment by zdenko** [comment]
@GeoffreyBooth the last commit fixes the returned value, but I just found another issue, which I'll fix later today.
Sorry, I was too quick with the push.

**Comment by GeoffreyBooth** [comment]
No worries.

The tests we have at least prove that this syntax compiles; it would be nice to also somehow prove that `first` and `last` are the values we expect them to be.

**Comment by zdenko** [comment]
@GeoffreyBooth the "issue" I noticed before is a false alarm.
When I checked this compiled output,
```coffeescript
foo = (list) ->
  ret =
    if list.length > 0
      [first, ..., last] = list
      [first, last]
    else
      []

###
foo = function(list) {
  var first, last, ret;
  return ret = list.length > 0 ? (([first] = list, [last] = slice.call(list, -1), list), [first, last]) : [];
};
###
```

I thought that this part `(([first] = list, [last] = slice.call(list, -1), list)` shouldn't include `list` at the end.
Although it's not a bug, since the returned value is correct, it would be nicer if it could be omitted from such expressions. 
It's probably a task for some other PR.

**Pull request merged**